aws/resource_aws_s3_bucket_object.go

```diff
@@ -377,47 +411,70 @@ func resourceAwsS3BucketObjectUpdate(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("error setting S3 object tags: %s", err)
 	}
 
+	if d.HasChange("legal_hold") {
+		_, err := conn.PutObjectLegalHold(&s3.PutObjectLegalHoldInput{
+			Bucket:    aws.String(d.Get("bucket").(string)),
+			Key:       aws.String(d.Get("key").(string)),
+			LegalHold: expandS3ObjectLockLegalHold(d.Get("legal_hold").([]interface{})),
+		})
+		if err != nil {
+			return fmt.Errorf("error putting S3 object lock legal hold: %s", err)
+		}
+	}
+
 	return resourceAwsS3BucketObjectRead(d, meta)
 }
 
 func resourceAwsS3BucketObjectDelete(d *schema.ResourceData, meta interface{}) error {
-	s3conn := meta.(*AWSClient).s3conn
+	conn := meta.(*AWSClient).s3conn
 
 	bucket := d.Get("bucket").(string)
 	key := d.Get("key").(string)
+	force := d.Get("force_delete").(bool)
 
-	if _, ok := d.GetOk("version_id"); ok {
```

Comment:
Collapse the versioned and non-versioned branches. `VersionId` is `nil` for unversioned buckets.